[Serve] [Docs] Document cancellation in Ray Serve (#39120)
ray-project/ray: doc/source/serve/http-guide.md

```

(serve-handling-client-disconnects-http)=
### Handling client disconnects

When an HTTP client disconnects before receiving a response, Serve cancels its in-flight request:

- If the proxy hasn't yet sent the request to a replica, the request is simply dropped.
- If the request has been sent to a replica, Serve attempts to interrupt the replica and cancel the request. When the replica enters an `await` statement, Serve raises an `asyncio.CancelledError`. Handle this exception in a try-except block to customize your deployment's behavior when a request is cancelled:

```{literalinclude} doc_code/http_guide/http_guide.py
:start-after: __start_basic_disconnect__
:end-before: __end_basic_disconnect__
:language: python
```

If there are no remaining `await` statements in the deployment's code before the request completes, the replica processes the request as usual, sends the response back to the proxy, and the proxy discards the response. Use `await` statements for blocking operations in a deployment, so in-flight requests can be cancelled in the deployment without waiting for the blocking operation to complete.

> Serve implements cascading cancellations. If a deployment is `awaiting` a response from another Ray task, Ray actor, or Serve deployment, Serve raises the `asyncio.CancelledError` in the `awaiting` deployment as well as the downstream task, actor, or deployment. This cancellation cascades through any chain of tasks, actors, or deployments that are `awaiting` another task, actor, or deployment. To block a cancellation from cascading to a downstream task, actor, or deployment, use `asyncio.shield`:

Reviewer: the cancellation in downstreams is independent of whether or not you are `await`ing the response. it will happen to any downstream task/actor call you've made

Author: Oh I didn't know that. I'll change the wording.

Reviewer: Can be more concise:

> Cancellation will cascade to any downstream deployment handle, task, or actor calls that were spawned in the deployment request handling method. These can handle the `asyncio.CancelledError` in the same was as the ingress deployment.

And once again, `shield` isn't specific to cascading. Mention it above when describing how to handle the error

Reviewer: Oh, I think I was looking at the outdated version. But I think what I wrote in the quote block above should suffice.

Author: Sounds good, I'll switch tot he block you wrote. Does the updated explanation about `shield` make sense?

Serve provides a library of HTTP adapters to help you avoid boilerplate code. The [later section](serve-http-adapters) dives deeper into how these works.
```

(serve-request-cancellation-http)=
### Request cancellation

When processing a request takes longer than the [end-to-end timeout](serve-performance-e2e-timeout) or an HTTP client disconnects before receiving a response, Serve cancels the in-flight request:

> - If the proxy hasn't yet sent the request to a replica, the request is simply dropped.

Reviewer 2: ```suggestion
- If the proxy hasn't yet sent the request to a replica, Serve simply drops the request.
```

> - If the request has been sent to a replica, Serve attempts to interrupt the replica and cancel the request. The `asyncio.Task` running the handler on the replica will be cancelled, raising an `asyncio.CancelledError` the next time it enters an `await` statement. See [the asyncio docs](https://docs.python.org/3/library/asyncio-task.html#task-cancellation) for more info. Handle this exception in a try-except block to customize your deployment's behavior when a request is cancelled:

Reviewer 2: ```suggestion
- If the request has been sent to a replica, Serve attempts to interrupt the replica and cancel the request. The `asyncio.Task` running the handler on the replica is cancelled, raising an `asyncio.CancelledError` the next time it enters an `await` statement. See [the asyncio docs](https://docs.python.org/3/library/asyncio-task.html#task-cancellation) for more info. Handle this exception in a try-except block to customize your deployment's behavior when a request is cancelled:
```

```{literalinclude} doc_code/http_guide/http_guide.py
:start-after: __start_basic_disconnect__
:end-before: __end_basic_disconnect__
:language: python
```

> If there are no remaining `await` statements in the deployment's code before the request completes, the replica processes the request as usual, sends the response back to the proxy, and the proxy discards the response. Use `await` statements for blocking operations in a deployment, so in-flight requests can be cancelled in the deployment without waiting for the blocking operation to complete.

Reviewer 2: ```suggestion
If no `await` statements are remaining in the deployment's code before the request completes, the replica processes the request as usual, sends the response back to the proxy, and the proxy discards the response. Use `await` statements for blocking operations in a deployment, so Serve can cancel in-flight requests in the deployment without waiting for the blocking operation to complete.
```

Cancellation cascades to any downstream deployment handle, task, or actor calls that were spawned in the deployment's request-handling method. These can handle the `asyncio.CancelledError` in the same way as the ingress deployment.

To prevent an async call from being interrupted by `asyncio.CancelledError`, use `asyncio.shield()`:

```{literalinclude} doc_code/http_guide/disconnects.py
:start-after: __start_shielded_disconnect__
:end-before: __end_shielded_disconnect__
:language: python
```

> When the request is cancelled, a cancellation error is raised inside the `SnoringSleeper` deployment's `__call__()` method. However, the cancellation is not raised inside the `snore()` call, so `ZZZ` is printed even if the request is cancelled. Note that `asyncio.shield` cannot be used on a `ServeHandle` call.

Reviewer: ```suggestion
When the request is cancelled, a cancellation error is raised inside the `SnoringSleeper` deployment's `__call__()` method. However, the cancellation is not raised inside the `snore()` call, so `ZZZ` is printed even if the request is cancelled. Note that `asyncio.shield` cannot be used on a `ServeHandle` call to prevent the downstream handler from being cancelled. You need to explicitly handle the cancellation error in that handle as well.
```

(ServeReplica:default_StreamingResponder pid=41052) INFO 2023-05-25 10:49:52,230 default_StreamingResponder default_StreamingResponder#qlZFCa yomKnJifNJ / default replica.py:634 - __CALL__ OK 1017.6ms
```

### Terminating the stream when a client disconnects

In some cases, you may want to cease processing a request when the client disconnects before the full stream has been returned.

> If you pass an async generator to `StreamingResponse`, it will be cancelled and raise an `asyncio.CancelledError` when the client disconnects.

Reviewer 2: ```suggestion
If you pass an async generator to `StreamingResponse`, it is cancelled and raises an `asyncio.CancelledError` when the client disconnects.
```
